numpy/core/overrides.py

```diff
@@ -187,25 +175,15 @@ def decorator(implementation):
         if docs_from_dispatcher:
             add_docstring(implementation, dispatcher.__doc__)
 
-        # Equivalently, we could define this function directly instead of using
-        # exec. This version has the advantage of giving the helper function a
-        # more interpettable name. Otherwise, the original function does not
-        # show up at all in many cases, e.g., if it's written in C or if the
-        # dispatcher gets an invalid keyword argument.
-        source = _wrapped_func_source.format(name=implementation.__name__)
-
-        source_object = compile(
-            source, filename='<__array_function__ internals>', mode='exec')
-        scope = {
-            'implementation': implementation,
-            'dispatcher': dispatcher,
-            'functools': functools,
-            'implement_array_function': implement_array_function,
-        }
-        exec(source_object, scope)
-
-        public_api = scope[implementation.__name__]
+        @functools.wraps(implementation)
+        def public_api(*args, **kwargs):
+            relevant_args = dispatcher(*args, **kwargs)
+            return implement_array_function(
+                implementation, public_api, relevant_args, args, kwargs)
 
+        public_api.__code__ = public_api.__code__.replace(
+                                        co_name=implementation.__name__,
+                                        co_filename='<string>')
```

Comment:
```suggestion
                                        co_filename='<__array_function__ internals>')
```
unless there's a reason not to.

Reply:
(Not sure what old behavior you're referring to, but making the string more specific sounds good.)

Reply:
> Not sure what old behavior you're referring to

The old code also passed the "filename": `compile(source, filename='<__array_function__ internals>', mode='exec')` printing `np.concatenate.__code__` as:
```
# before:
<code object concatenate at 0x7f4294053450, file "<__array_function__ internals>", line 2>
# with the change:
<code object concatenate at 0x7f6370198a80, file "<string>", line 178>
```

Github had some CI hiccups, but CI fails on Python 3.7.  So we need a workaround or keep the old code as well.  But it does give a 15% performance boost on my machine, so that seems worthwhile even if we add a switch (that can be retired very soon):
```
       before           after         ratio
     [c4e20ec4]       [14dcc3ad]
     <main>           <no_compile>
-      71.1±0.6ms       61.3±0.3ms     0.86  bench_import.Import.time_numpy
```